node/core/dispute-coordinator/src/lib.rs

```diff
@@ -0,0 +1,959 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Implements the dispute coordinator subsystem.
+//!
+//! This is the central subsystem of the node-side components which participate in disputes.
+//! This subsystem wraps a database which tracks all statements observed by all validators over some window of sessions.
+//! Votes older than this session window are pruned.
+//!
+//! This subsystem will be the point which produce dispute votes, either positive or negative, based on locally-observed
+//! validation results as well as a sink for votes received by other subsystems. When importing a dispute vote from
+//! another node, this will trigger the dispute participation subsystem to recover and validate the block and call
+//! back to this subsystem.
+
+use std::sync::Arc;
+
+use polkadot_node_primitives::{CandidateVotes, SignedDisputeStatement};
+use polkadot_node_subsystem::{
+	messages::{
+		DisputeCoordinatorMessage, ChainApiMessage,
+	},
+	Subsystem, SubsystemContext, FromOverseer, OverseerSignal, SpawnedSubsystem,
+	SubsystemError,
+	errors::{ChainApiError, RuntimeApiError},
+};
+use polkadot_node_subsystem_util::rolling_session_window::{
+	RollingSessionWindow, SessionWindowUpdate,
+};
+use polkadot_primitives::v1::{
+	SessionIndex, CandidateHash, Hash, CandidateReceipt, DisputeStatement, ValidatorIndex,
+	ValidatorSignature,
+};
+
+use futures::prelude::*;
+use futures::channel::oneshot;
+use kvdb::KeyValueDB;
+use parity_scale_codec::Error as CodecError;
+
+use db::v1::ActiveDisputes;
+
+mod db;
+
+const LOG_TARGET: &str = "parachain::dispute-coordinator";
+
+// It would be nice to draw this from the chain state, but we have no tools for it right now.
+// On Polkadot this is 1 day, and on Kusama it's 6 hours.
+const DISPUTE_WINDOW: SessionIndex = 6;
```

Comment:
> but we have no tools for it right now

could you elaborate why can't we use `SessionInfo::dispute_period` from the runtime like we did previously?

Reply:
There is no `dispute_period` in `struct SessionInfo`. That is exactly what I mean by 'no tools for it' - we'd need to change the type definition or add a new runtime API.

Reply:
Sorry, I meant in HostConfiguration.

Reply:
But we can probably rely on the fact that session_info will store sessions based on `dispute_period`?

Reply:
https://github.com/paritytech/polkadot/issues/3227 <- We need to expose it, and that's beyond scope here IMO